src/System.Memory/src/System/MemoryExtensions.Fast.cs

```diff
@@ -101,5 +103,120 @@ public static partial class MemoryExtensions
+                    // is 1/2/4 byte-aligned but not also 8-byte aligned. Unfortunately, to catch that case we would have to walk the 
+                    // struct element sizes to determine the actual struct alignment.
+
+                    // This check also returns false if it's coincidentally aligned to TTo in the case where sizeof(TFrom)<sizeof(TTo). This is
+                    // because the source location can move such that it no longer adheres to the alignment requirement of TTo while still
+                    // following the alignment rules for it's declared/constructed type, TFrom.
+                    void* pointer = Unsafe.AsPointer(ref MemoryMarshal.GetReference(source));
+
+                    // First we have to make sure the given pointer follows the alignment rules of its TFrom type. Since the TFrom alignment will
+                    // be preserved even if the memory is moved, we can safely make assumptions about the alignment of TTo without the need for pinning.
+                    int sizeOfTFrom = Math.Min(8, Unsafe.SizeOf<TFrom>());
+                    if (new IntPtr(pointer).ToInt64() % sizeOfTFrom == 0)
+                    {
+                        if (sizeOfTFrom == 8 || Math.Min(8, Unsafe.SizeOf<TTo>()) <= sizeOfTFrom)
+                        {
+                            output = MemoryMarshal.Cast<TFrom, TTo>(source);
+                            return true;
+                        }
+                    }
+                    output = null;
+                    return false;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Attempts to cast a Span of one primitive type <typeparamref name="TFrom"/> to another primitive type <typeparamref name="TTo"/>.
+        /// </summary>
+        /// <param name="source">The source slice, of type <typeparamref name="TFrom"/>.</param>
+        /// <param name="output">The destination  of type <typeparamref name="TTo"/>.</param>
+        /// <remarks>If <typeparamref name="TTo"/> is 8-byte aligned and <paramref name="source"/> points to a valid aligned <typeparamref name="TTo"/> address,
+        /// this will always return true. This is because all C# primitives are aligned against at most 8 bytes, so a pointer that is 8 byte aligned 
+        /// will be aligned to all primitives.</remarks>
+        /// <remarks>If <paramref name="source"/> doesn't point to an address that follows the os-specific alignment rules of <typeparamref name="TFrom"/>, then 
+        /// this will always return false</remarks>
+        /// <remarks>Does not require the memory being spanned over to be fixed.</remarks>
+        /// <returns>True if successful; else False</returns>
+        public static bool TryCast<TFrom, TTo>(this Span<TFrom> source, out Span<TTo> output) where TFrom : struct where TTo : struct
```

Comment:
The Fast and Portable implementations of this API are essentially identical. Should we just have one copy in [MemoryExtensions.cs](https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/MemoryExtensions.cs)? MemoryExtensions.Fast.cs only contains extension methods with different implementation and calls the Span class (within CoreLib) directly.

Reply:
I expect that we will move MemoryExtensions to CoreLib. Once that happens we can have much better, faster and accurate, platform-specific implementation of this API. I do not think we need to bother with sharing it.

Reply:
They're duplicated so we can replace the Fast implementation once it gets moved to corelib. 
From the commit comments:

>Implemented for both the fast and portable implementations in equivalent ways, though eventually the fast implementation should be moved to corelib with the other MemoryExtensions.Fast methods and implemented using compiler directives (e.g. #IF intelProcessorCompiled) or maybe an intrinsic for the else case.

Reply:
Right, this is hard to get right in portable way. I would suggest that the portable implementation just checks that `sizeof(TTo) <= sizeof(IntPtr)` and returns false otherwise.

Reply:
Maybe even do it for x86/x64 and avoid dependency on the RuntimeInformation package.